platforms/extensibility/plugin-use/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequestApplicator.java

```diff
@@ -87,136 +81,57 @@ public DefaultPluginRequestApplicator(
-                    if (pluginVersion != null) {
-                        pluginVersionTracker.setPluginVersionAt(
-                            classLoaderScope,
-                            result.found.getPluginId().getId(),
-                            pluginVersion
-                        );
-                    }
-                }
-            });
-        }
-    }
-
-    private void applyPlugin(PluginManagerInternal target, Result result, PluginImplementation<?> impl) {
-        applyPlugin(result.request, result.found.getPluginId(), () -> {
-            if (result.request.isApply()) {
-                target.apply(impl);
+            if (request.isApply()) {
+                pluginApplyActions.add(new ApplyAction(request, resolved));
             }
-        });
-    }
 
-    // We're making an assumption here that the target's plugin registry is backed classLoaderScope.
-    // Because we are only build.gradle files right now, this holds.
-    // It won't for arbitrary scripts though.
-    private void applyLegacyPlugin(PluginManagerInternal target, Result result, PluginId id) {
-        applyPlugin(result.request, id, () -> {
-            if (result.request.isApply()) {
-                target.apply(id.toString());
+            String pluginVersion = resolved.getPluginVersion();
+            if (pluginVersion != null) {
+                pluginVersionTracker.setPluginVersionAt(
+                    classLoaderScope,
+                    resolved.getPluginId().getId(),
+                    pluginVersion
+                );
             }
-        });
-    }
-
-    private List<Result> resolvePluginRequests(PluginRequests requests, PluginResolver effectivePluginResolver) {
-        return collect(requests, request -> {
-            PluginRequestInternal configuredRequest = pluginResolutionStrategy.applyTo(request);
-            return resolveToFoundResult(effectivePluginResolver, configuredRequest);
-        }).stream()
-            .filter(r -> !r.alreadyApplied)
-            .collect(Collectors.toList());
-    }
-
-    private static void addPluginArtifactRepositories(PluginArtifactRepositories resolveContext, RepositoryHandler repositories) {
-        resolveContext.applyRepositoriesTo(repositories);
-    }
+        }
 
-    private static void defineScriptHandlerClassScope(ScriptHandlerInternal scriptHandler, ClassLoaderScope classLoaderScope, Iterable<PluginImplementation<?>> pluginsFromOtherLoaders) {
-        exportBuildLogicClassPathTo(classLoaderScope, scriptHandler.getInstrumentedScriptClassPath());
-        for (PluginImplementation<?> pluginImplementation : pluginsFromOtherLoaders) {
-            classLoaderScope.export(pluginImplementation.asClass().getClassLoader());
+        // Configure the resolution
+        dependencyVisitor.getAdditionalDependencies().forEach(scriptHandler::addScriptClassPathDependency);
+        if (!dependencyVisitor.getSubstitutions().isEmpty()) {
```

Comment:
Do some testing to verify what happens when old:coords:required and new:coords:preferred do a module replacement conflict resolution. What version is selected? What happens if there is overlap between versions? What coordinates are selected? What if the old coodinates do not have the version that the new coordinates have?